### PR TITLE
Refactor Functions runtime delegation process to no longer depend on cli options and context

### PR DESCRIPTION
**Pull request by taeold**
Code for parsing CF3 triggers from function source relies on CLI options/context object to retrieve active project id, functions source directory, project root, etc.

The refactoring here pulls out all references to CLI options and context object from the runtime delegate caller. This is done in preparation for having another caller for the runtime delegate (the Functions Emulator) in the near future.

### CODE REVIEW

**Review by colerogers — APPROVED** [review]
Overall this looks like a clean refactor change, lgtm

File: src/deploy/functions/prepare.ts [hunk 1]
```diff
@@ -53,18 +54,28 @@ export async function prepare(
   options: Options,
   payload: args.Payload
 ): Promise<void> {
-  if (!options.config.src.functions) {
-    return;
+  const projectId = needProjectId(options);
+
+  const sourceDirName = options.config.get("functions.source") as string;
+  if (!sourceDirName) {
+    throw new FirebaseError(
```
Comment (inlined) [inline src/deploy/functions/prepare.ts:61]: Be careful here. I think I tried this and we started throwing exceptions for projects that didn't include functions.

File: src/deploy/functions/runtimes/index.ts [hunk 2]
```diff
@@ -100,36 +98,34 @@ export interface RuntimeDelegate {
...
 }
 
-type Factory = (context: args.Context, options: Options) => Promise<RuntimeDelegate | undefined>;
+export interface DelegateContext {
+  projectId: string;
+  // Absolute path of the Firebase project directory.
+  projectDir: string;
+  // Absolute path of the source directory.
+  sourceDir: string;
+  runtime: string;
```
Comment (inlined) [inline src/deploy/functions/runtimes/index.ts:107]: nit: I might call this a runtime override or something. This overrules the normal sniffing we do.

Comment (inlined) [reply src/deploy/functions/runtimes/index.ts:107]: Also, shouldn't it be optional since the firebase.json value is optional?

**Review by inlined — COMMENTED** [review]
Looking good. Sorry I've been AWOL on reviews.

**Pull request merged**